### PR TITLE
[Notebook] Open Kernel Alias (Kusto) notebook with Kernel Alias (Kusto) as selected kernel

### LINKED ISSUE #13737: Kusto Notebooks Show SQL as Kernel on Reopen
- Azure Data Studio Version:
Version: 1.25.0-rc1 (user setup)
Commit: 634ea0ab6a60543d712dbc77266059de1d70c08c
Date: 2020-12-08T19:09:44.467Z
VS Code: 1.48.0
Electron: 9.3.0
Chrome: 83.0.4103.122
Node.js: 12.14.1
V8: 8.3.110.13-electron.0
OS: Windows_NT x64 10.0.19042

Steps to Reproduce:

1. Create a new notebook connecting to ~Python~ KQL Kernel. 
2. Save the file.
3. Reopen it. The kernel will not be set to ~Python~ KQL and will get changed back to the default, which is SQL.

An additional ask would be to remember the default connection string / compute location, where applicable.

### PR DESCRIPTION
**Pull request by VasuBhog**
This PR fixes #13737.

The scenarios covered: 
- Opens kernel alias Notebook properly with kernel set to kernel alias.
- Switches Kernels properly if kernel alias notebook is open first.
- If Kusto Extension is not installed it will open notebook with SQL kernel.

### CODE REVIEW

File: src/sql/workbench/services/notebook/browser/models/notebookModel.ts [hunk 1]
```diff
@@ -951,6 +951,11 @@ export class NotebookModel extends Disposable implements INotebookModel {
 
 	private async updateKernelInfoOnKernelChange(kernel: nb.IKernel, kernelAlias?: string) {
 		await this.updateKernelInfo(kernel);
+		this.kernelAliases.forEach(kernel => {
```
Comment (chlafreniere) [inline src/sql/workbench/services/notebook/browser/models/notebookModel.ts:954]: Tests 😄 ?

**Review by barbaravaldez — APPROVED** [review]

**Comment by Charles-Gagnon** [comment]
@VasuBhog Did you mean to link the issue to be closed? addresses isn't a valid keyword for that if so

**Comment by VasuBhog** [comment]
> @VasuBhog Did you mean to link the issue to be closed? addresses isn't a valid keyword for that if so

Oh yeah, I became quite used to ask mode bugs and format. I'll update use fixes.

**Comment by chlafreniere** [comment]
> > @VasuBhog Did you mean to link the issue to be closed? addresses isn't a valid keyword for that if so
> 
> Oh yeah, I became quite used to ask mode bugs and format. I'll update use fixes.

I would be very happy if folks forgot the ask mode process because it was a rare occurrence 😉

**Comment by VasuBhog** [comment]
This change does not require us to call the changeKernel emitter since we are able to pass the appropriate kernel to NotebookActions via the `updateKernelInfoOnKernelChange`.
This way we can directly change the kernel via the languageInfo on load which is both efficient and keeps languageInfo on NotebookModel side as expected.

**Comment by VasuBhog** [comment]
Will be adding more tests to NotebookActions and a possible smoke test in another PR.

File: src/sql/workbench/contrib/notebook/test/electron-browser/notebookModel.test.ts [hunk 2]
```diff
@@ -865,9 +886,47 @@ suite('notebook model', function (): void {
 		assert.equal(output.metadata['multi_connection_mode'], true, 'multi_connection_mode not saved correctly to notebook metadata');
 	});
 
-	async function loadModelAndStartClientSession(): Promise<NotebookModel> {
+	test('Should not language info kernel alias name even if kernel spec is SQL', async function () {
```
Comment (lucyzhang929) [inline src/sql/workbench/contrib/notebook/test/electron-browser/notebookModel.test.ts:889]: description is not clear - might be missing a word or two?

File: src/sql/workbench/contrib/notebook/test/electron-browser/notebookModel.test.ts [hunk 3]
```diff
@@ -865,9 +886,47 @@ suite('notebook model', function (): void {
...
-		mockContentManager.setup(c => c.loadContent()).returns(() => Promise.resolve(expectedNotebookContent));
+		mockContentManager.setup(c => c.loadContent()).returns(() => Promise.resolve(expectedKernelAliasNotebookContentOneCell));
+		defaultModelOptions.contentManager = mockContentManager.object;
+
+		queryConnectionService.setup(c => c.getActiveConnections(TypeMoq.It.isAny())).returns(() => null);
+
+		// Given I have a session that fails to start
+		sessionReady.resolve();
+
+		let model = new NotebookModel(defaultModelOptions, undefined, logService, undefined, new NullAdsTelemetryService(), queryConnectionService.object, configurationService);
```
Comment (lucyzhang929) [inline src/sql/workbench/contrib/notebook/test/electron-browser/notebookModel.test.ts:899]: can we do `let model = loadModelAndStartclientSession(expectedKernelAliasNotebookContentOneCell);` here?

**Review by lucyzhang929 — COMMENTED** [review]
I followed these steps:
1. Created new notebook
2. Changed kernel to Kusto
3. Saved notebook
4. Reopened notebook

The kernel drop down still showed SQL.
Am I missing something? I have the Kusto extension installed.

**Comment by VasuBhog** [comment]
> I followed these steps:
> 
> 1. Created new notebook
> 2. Changed kernel to Kusto
> 3. Saved notebook
> 4. Reopened notebook
> 
> The kernel drop down still showed SQL.
> Am I missing something? I have the Kusto extension installed.

So interesting enough. This is a regression. I just used the December release of ADS and found that language_info metadata gets updated properly with Kusto as the name for Kusto Notebooks. However, due to some new changes that must have gone in recently, the current stable and insiders do not save notebook language info properly for Kusto notebooks.
I will open a new issue for that to be tracked.

Metadata for Kusto notebooks should be like this:
```
"metadata": {
        "kernelspec": {
            "name": "SQL",
            "display_name": "SQL",
            "language": "sql"
        },
        "language_info": {
            "name": "kusto",
            "version": ""
        }
    }
```

Any Kusto Notebook created with the December release at the moment will work properly with this change, you can manually update as well for now to test the functionality.

**Comment by VasuBhog** [comment]
> I followed these steps:
> 
> 1. Created new notebook
> 2. Changed kernel to Kusto
> 3. Saved notebook
> 4. Reopened notebook
> 
> The kernel drop down still showed SQL.
> Am I missing something? I have the Kusto extension installed.

Just opened this issue: #14545
I was working with Notebooks that had language_info already set to Kusto or notebooks that had already connected to a Kusto connection. 
Will look into updating the language_info when switching kernels if that is what we expect.

File: src/sql/workbench/contrib/notebook/test/electron-browser/notebookModel.test.ts (hunk 3 above)
Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/test/electron-browser/notebookModel.test.ts:899]: So this was one of the issues I ran into. I was trying to do this but couldn't due to multiple errors.

We need the model.loadContents() in order for the metadata to properly set to the language_info to kusto and not the defaultKernel.

But there is a further issue if I use the function you state and the loadContents() as I get `TypeError: Cannot read property 'createInstance' of undefined` on [loadContents method](https://github.com/microsoft/azuredatastudio/blob/main/src/sql/workbench/services/notebook/browser/models/notebookModel.ts#L412). I'll look into this further on Test Mondays, to add more tests and do more revamping incrementally.

**Review by lucyzhang929 — APPROVED** [review]

**Pull request merged**